content/languages/c/authoring/memory-management-techniques.md

```diff
@@ -6,85 +6,129 @@ sidebar: "language:c"
 
 <details>
+    <summary>Example</summary>
 
-Solution:
+Preloaded:
 
 ```c
-//get all prime numbers less than upto
-int* get_primes(int upto, int* size) {
+//Provide a typedef for constants.
+//If you really want to use strings for some reason, you can use
+//constants of type const char*, but it's recommended to take
+//this step even further and use an enum.
+typedef const char * const Player;
+
+//define constants
+Player BLACK = "BLACK";
+Player WHITE = "WHITE";
+Player NONE  = "NONE";
+```
 
-    //the solution allocates required memory
-    int* result = malloc(sizeof(int) * ...);
+Solution:
 
-    //... fill result with primes
-    //...
+```c
 
-    *size = ...; //assign amount of primes
-    return result;
+//Since Codewars does not allow header files for kata, declarations need to be repeated
+typedef const char * const Player;
+extern Player BLACK;
+extern Player WHITE;
+extern Player NONE; 
+
+
+Player who_won(const char* board) //typedef used for return type 
+{
+    if(...) {
+      return BLACK;    //return constant instead of an allocated string
+    } else if (...) {
+      return WHITE;
+    } else {
+      return NONE;
+    }
 }
 ```
 
-Test suite:
+
+Tests:
 
 ```c
-Test(fixed_tests, should_return_2_and_3_for_4) {
-    
-    int expected[] {2, 3}, expected_size = 2;
-    int actual_size;
-    
-    //call user solution and expect it to allocate the returned array
-    int* actual = get_primes(4, &actual_size);
+//Since Codewars does not allow header files for kata, declarations need to be repeated
+typedef const char * const Player;
+extern Player BLACK;
+extern Player WHITE;
+extern Player NONE; 
 
-    //...assert on actual_size
-    //...assert on contents of actual
+Player who_won(const char* board);
 
-    //after performing all necessary assertions,
-    //free the array allocated by the user solution
-    free(actual);
+Test(fixed_tests, no_one_won) {
+ 
+  Player winner = who_won("B");
+
+  //constants can be asserted on with cr_assert_eq
+  cr_assert_eq(winner, NONE, "Expected: [%s], but was: [%s]", NONE, winner);
```

Comment:
Non-string assertion... Although it's supposed to be the same pointer, I think if it's not an enum for whatever reason, it should be treated as a string. Someone will return their own string literal eventually just to get rid of extra declarations in the code.

Reply:
> Someone will return their own string literal eventually just to get rid of extra declarations in the code.

Fair point.

Reply:
I removed constant strings completely, replaced them with enums.